### PR TITLE
New openapi v3 testing file client

### PR DESCRIPTION
**Pull request by seans3**
* Creates testing openapi file client, which supports hard-coded OpenAPI V3 GroupVersion specification files in the `testdata` subdirectory.
* Adds unit tests for new testing client and groupversion.
* Refactors `kubectl` `explain_test` and `generator_test` to use this new testing file client.
* This new file client will be used to test another upcoming OpenAPI V3 client.

Retry of: https://github.com/kubernetes/kubernetes/pull/115465

This previous PR broke the publishing bot because of a cross-module filepath (there are no presubmits for this issue): xref https://github.com/kubernetes/kubernetes/issues/56876#issuecomment-1430665316 .

The current PR embeds the testing files directly using `//go:embed` directive, so there are no longer filepaths.

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

**Comment by apelisse** [comment]
Can you link to the previous PR, explain what happened, and describe what's different?

**Comment by seans3** [comment]
> Can you link to the previous PR, explain what happened, and describe what's different?

Updated in PR description

File: staging/src/k8s.io/client-go/openapi/openapitest/fileclient.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+// NewFileClient returns a pointer to a testing
+// FileOpenAPIClient, which parses the OpenAPI GroupVersion
+// files in the passed directory path. Example GroupVersion
+// OpenAPI filename for apps/v1 would look like:
+//
+//	apis__apps__v1_openapi.json
+//
+// An example to invoke this function for the test files:
+//
+//	NewFileClient("../../../testdata/openapi/v3")
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/openapi/openapitest/fileclient.go:39]: this still seems too fragile ... can we put the test fixtures in a testing-only package and not allow any backsteps at all?

**Comment by liggitt** [comment]
the json files are *really* big... can we copy a subset of them for testing instead of all of them (maybe something in the core group, and a couple versions outside the core group)?

**Comment by liggitt** [comment]
> can we copy a subset of them for testing instead of all of them (maybe something in the core group, and a couple versions outside the core group)?

alternately, relocate the existing .json files to `staging/src/k8s.io/client-go/openapi/openapitest/testdata` and use https://pkg.go.dev/embed to make an unconfigurable `k8s.io/client-go/openapi/openapitest#NewTestClient(t *testing.T) openapi.Client` method that surfaces them

**Comment by liggitt** [comment]
> alternately, relocate the existing .json files to `staging/src/k8s.io/client-go/openapi/openapitest/testdata` and use https://pkg.go.dev/embed to make an unconfigurable `k8s.io/client-go/openapi/openapitest#NewTestClient(t *testing.T) openapi.Client` method that surfaces them

something along the lines of https://github.com/liggitt/kubernetes/commits/openapi-test

**Comment by apelisse** [comment]
I like where the files live right now. And I like the idea of the go-embed too, can we have one copy? Git doesn't hold two copies of the files in the repo since the hashes are equal, so technically it's really not that bad ...

**Comment by liggitt** [comment]
> I like where the files live right now.

you can't reference things in k8s.io/kubernetes from inside staging modules... I'm unsure of the goals for this change ... are we happy with a subset or would we want this to stay up to date as our generated spec files change?

**Comment by apelisse** [comment]
I think I like that we have an easy reference to the current openapi in `api/openapi-spec/v3`, so I wouldn't really want to move the files into a deeply nested directory.

I think for that test specifically, we absolutely don't need a big set of files, so we could keep a couple of small one and be happy (we don't really look at the content at all). Would that work?

**Comment by seans3** [comment]
> > I like where the files live right now.
> 
> you can't reference things in k8s.io/kubernetes from inside staging modules... I'm unsure of the goals for this change ... are we happy with a subset or would we want this to stay up to date as our generated spec files change?

We would be happy with a subset...it does not need to be the most up-to-date generated spec.

**Comment by apelisse** [comment]
Unrelated test-failure
/retest

**Pull request merged**